### PR TITLE
Browse by Controlled Vocabulary Page

### PR DESCRIPTION
**Pull request by nona-luypaert**
## References
Rest PR: https://github.com/DSpace/DSpace/pull/8826
Contract PR: https://github.com/DSpace/RestContract/pull/225

## Description
This PR adds a browse page for controlled vocabularies. Which vocabulary to use is dynamically configurable from the backend.

## Instructions for Reviewers
**Changes**
- Took modal-logic out of VocabularyTreeview and put it in its own component: VocabularyTreeviewModal. (DynamicOnebox and SearchHierarchyFilter now call this component.)
- Refactored VocabularyTreeviewNodeModel, VocabularyTreeviewComponent and VocabularyTreeviewService to keep track of currently selected items and the `isSelected`-state of a checkbox.
- Added a `multiSelect`-input to VocabularyTreeviewComponent to control whether checkboxes should be displayed or not. Also added an extra output to emit deselecting-events.
- Added new BrowseDefinition models which correspond to the new `browseTypes` on the REST-side (valueList, flatBrowse and hierarchicalBrowse).
- Added new BrowseDefinitionRestRequst, BrowseResponseParsingService and BrowseDefinitionFindAllDataImpl to be able to retrieve data from the `browses` REST-endpoint.
- Added a (themed) BrowseByTaxonomyPage, which displays a VocabularyTreeview, allows the user to select/deselect multiple items and then redirects to the search-page with the selected items as search filters.
- Added an extra "Browse by Subject Category"-link to "All of DSpace". (This is automatically generated.)
- Added new tests for VocabularyTreeview, BrowseByTaxonomyPage and BrowseResponseParsingService, fixed other tests where necessary.  

**Review instructions**
- Verify the other browse pages still work as before.
- Go to the new browse page ("By Subject Category" under "All of DSpace")
- Verify the search bar and reset button work as expected.
- Select a couple of subjects, click "Browse" and verify that you are redirected to the search-page. Verify that your chosen subjects show up in the URL and are checked in the Subject sidebar facet.
- Select and deselect a couple of subjects. After clicking "Browse", verify that the subjects you deselected don't show up in the URL and aren't checked in the Subject sidebar facet.
- Verify the original use cases of VocubalaryTreeview:
    - Verify the modal works as before when choosing a subject during item submission (**config**: [input-type of subject](https://github.com/DSpace/DSpace/blob/main/dspace/config/submission-forms.xml#L197) in backend should be set to `onebox`)
    - Verify the modal works as before when choosing a subject in the Subject sidebar facet. (**config**: make sure to enable [vocabularies](https://github.com/DSpace/dspace-angular/blob/main/config/config.example.yml#L368) in config.yml)

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: src/app/core/shared/hierarchical-browse-definition.model.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  get self(): string {
+    return this._links.self.href;
+  }
+
+  @deserialize
+  _links: {
+    self: HALLink;
+    entries: HALLink;
+    items: HALLink;
```
Comment (tdonohue) [inline src/app/core/shared/hierarchical-browse-definition.model.ts:40]: `entries` and `items` are not valid for this type of Browse.

File: src/app/core/shared/non-hierarchical-browse-definition.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  @autoserializeAs('metadata')
+  metadataKeys: string[];
+
+  @autoserialize
+  dataType: BrowseByDataType;
+
+  @deserialize
+  _links: {
+    self: HALLink;
+    entries: HALLink;
```
Comment (tdonohue) [inline src/app/core/shared/non-hierarchical-browse-definition.ts:29]: `entries` is only valid if the type is `valueList`.

File: src/app/core/shared/non-hierarchical-browse-definition.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    items: HALLink;
```
Comment (tdonohue) [inline src/app/core/shared/non-hierarchical-browse-definition.ts:30]: `items` is only valid if the type is `flatBrowse`.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@nona-luypaert : The basic functionality works, but I'm seeing some errors reported in my  browser's DevTools console for each of the browse by endpoints:

Here's what I'm doing:
1. First, I'm using this in production mode (`yarn build:prod; yarn serve:ssr`).
2. Go to homepage & open up browser DevTools (to watch for errors)
3. As an anonymous user, visit "All of DSpace -> By Subject Category`
    * **Error** appears in console. I see an odd request to `/server/api/dso/find?uuid=undefined` which returns a 400.  Not sure why that endpoint is being accessed here?
4. Click on `All of DSpace -> By Author` (or `By Subject`)
    * **Error** appears in console, as the UI tries to access `/items` for these `valueList` browses, and that's invalid.
5. Click on `All of DSpace -> By Title` (or `By Date`)
    * **Error** appears in console, as the UI tries to access `/entries` for these `flatBrowse` browses, and that's invalid.

(NOTE: I believe many of these errors above may be caused by the backend returning invalid `_link` values based on the browse type.  So, I've noted [this feedback also on that PR](https://github.com/DSpace/DSpace/pull/8826#pullrequestreview-1442727423).)

Finally, a tiny design issue (no fault of this PR, but this PR makes it more evident).  The open/close arrows in the list are very large.  They appear to be using `class="fa-2x"` (i.e. 2 times the size) instead of simply `class="fa"`.  I'd recommend we change these to be the latter, as that'd make them the same size as the same arrows on the `/community-list` page of the UI.  (NOTE: This isn't a fault of this PR though, so I won't require it.  I just think it'd make the design look slightly better.)
arrows-large

Overall, this PR looks good and works.  I've verified it doesn't appear to impact any other browse by features, and that the Hierarchical browse option on the search page & submission page both appear to still work.

Finally, I should note this PR is currently failing a lot of our e2e tests.  I'm not sure why it is failing, but we will need to determine a proper fix obviously.

**Comment by artlowel** [comment]
Thanks for the review @tdonohue!

All the console errors you're seeing also happen on demo; they're not caused by this PR.

After looking into it I expect the ones with the incorrect `/items` and `/entries` links will be fixed once they're removed on the rest side. Nona will remove them from the models in the UI, but even after she does, they'll still be added if they're part of the rest response. Typescript is basically only adding info for us and the compiler to make sense of those responses, but doesn't prevent properties that aren't part of the model being added. Since the UI checks for the presence of the links before calling them, the calls should stop if rest stops adding those links.

The `/server/api/dso/find?uuid=undefined` call is [caused by DSOEditMenuResolver](https://github.com/DSpace/dspace-angular/blob/d2fa8cda6ac2c0139ebf9182140eda45192a1898/src/app/shared/dso-page/dso-edit-menu.resolver.ts#L53), which assumes it's on a community or collection homepage and tries to add the context sensitive dso edit menu. We can fix that in this PR

**Review by tdonohue — CHANGES_REQUESTED** [review]
@nona-luypaert and @artlowel :  This is looking great overall.  I'm almost a +1, but I've found a few minor bugs.  Simply put, the ~~"Search" and~~ "Reset" buttons under "Browse by Subject Category" don't work.

vocab_buttons

* ~~If I enter in text and click "Search", nothing happens. The button doesn't do anything.  Maybe it shouldn't be there?  Or maybe it should do the same thing as the "Browse" button?~~
    * Disregard this. I was misunderstanding its purpose. It *does* work.  I was entering text and pressing Enter to perform a search. I now see that I can also enter text and click "Search".  My mistake :) 
* If I enter in text, select a vocabulary value, and click "Reset".  The form will "appear" to reset.  But, if you then search for the **same** vocabulary value again, you'll see it's still selected (see the screenshot above where the value appears in green). 
    * Similarly, if you select a value, click "Reset" and *then* click "Browse"... you'll get results for the selected value. 
    * My assumption is that "Reset" should cause any selected values to be deselected.

Beyond that, everything else is working well now & the updates look good.  I've also verified that the Browse by menu doesn't appear when SRSC vocabulary is disabled in the submission-forms.xml (which is good!).  Also verified that the e2e tests pass locally when the backend PR is installed.

Once these buttons are fixed, I'm a +1

**UPDATE: Found a possible SSR issue as well. See next comment.**

**Comment by tdonohue** [comment]
UPDATE: I've also found what appears to be a server-side rendering issue?

* Start the site in production mode (`yarn build:prod; yarn serve:ssr`).  Keep an eye on the console where you started it from.
* Visit http://localhost:4000/ 
* Go to the "Browse by Subject Category" page.
* Click reload in your browser window.
* Back in the console (where you ran `yarn serve:ssr`) you'll see this error:
```
GET /browse/srsc 200 1519.656 ms - -
ERROR Error [NullInjectorError]: R3InjectorError(BrowseByPageModule)[VocabularyTreeviewService -> VocabularyTreeviewService -> VocabularyTreeviewService -> VocabularyTreeviewService]:
  NullInjectorError: No provider for VocabularyTreeviewService!
    at NullInjector.get (C:\dspace-angular\dist\server\main.js:1:5375951)
    at R3Injector.get (C:\dspace-angular\dist\server\main.js:1:5382668)
    at R3Injector.get (C:\dspace-angular\dist\server\main.js:1:5382668)
    at R3Injector.get (C:\dspace-angular\dist\server\main.js:1:5382668)
    at R3Injector.get (C:\dspace-angular\dist\server\main.js:1:5382668)
    at ChainedInjector.get (C:\dspace-angular\dist\server\main.js:1:5451112)
    at lookupTokenUsingModuleInjector (C:\dspace-angular\dist\server\main.js:1:5303022)
    at getOrCreateInjectable (C:\dspace-angular\dist\server\main.js:1:5303663)
    at Object.ɵɵdirectiveInject (C:\dspace-angular\dist\server\main.js:1:5417093)
    at NodeInjectorFactory.VocabularyTreeviewComponent_Factory [as factory] (C:\dspace-angular\dist\server\7377.js:1:658248) {
  ngTempTokenPath: null,
  ngTokenPath: [
    'VocabularyTreeviewService',
    'VocabularyTreeviewService',
    'VocabularyTreeviewService',
    'VocabularyTreeviewService'
  ]
}
```

**Review by tdonohue — CHANGES_REQUESTED** [review]
@nona-luypaert :  Tested again today.  It's *better*, but still seeing an odd bug with the "Reset" button.  The SSR issue is fixed though.

Here's the bug I'm seeing.. the "Reset" button is still not clearing out all previously selected data:
1. Visit "Browse by Subject Categories"
2. Select 2 categories.  They can even be top level categories, e.g. "Mathematics" and "Natural Sciences"
3. Click "Reset".  Those categories will appear deselected (good!)
4. Now, select a category you didn't previously select, e.g. "Humanities and Religion"
5. Click "Browse"
6. You'll go to the search page with **all three** categories added to the filter.  It appears the "Reset" button isn't clearing all the data.  I'd expect that you should only see the last filter (e.g. "Humanities and Religion").

Beyond that, everything else is working perfectly.  So, this is the final bug that I'm seeing.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @nona-luypaert !  This looks good now.  Retested today and everything is working as expected now.

**Pull request merged**